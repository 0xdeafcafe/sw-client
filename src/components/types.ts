export type DisplayProps = {
  currentRoot: string;
  item: FilmItemProps &
    PersonItemProps &
    PlanetItemProps &
    SpeciesItemProps &
    StarshipItemProps &
    VehicleItemProps;
};

export type FilmProps = {
  item: FilmItemProps;
  isItemFavourite: boolean;
  setIsItemFavourite: (boolean: boolean) => void;
};

export type FilmItemProps = {
  characters: string[];
  created: string;
  director: string;
  edited: string;
  episode_id: number;
  opening_crawl: string;
  planets: string[];
  producer: string;
  release_date: string;
  species: string[];
  starships: string[];
  title: string;
  url: string;
  vehicles: string[];
};

export type PersonProps = {
  item: PersonItemProps;
  isItemFavourite: boolean;
  setIsItemFavourite: (boolean: boolean) => void;
};

export type PersonItemProps = {
  created: string;
  edited: string;
  birth_year: string;
  eye_color: string;
  films: string[];
  gender: string;
  hair_color: string;
  height: string;
  homeworld: string;
  name: string;
  skin_color: string;
  species: string[];
  starships: string[];
  url: string;
  vehicles: string[];
  mass: string;
};

export type PlanetProps = {
  item: PlanetItemProps;
  isItemFavourite: boolean;
  setIsItemFavourite: (boolean: boolean) => void;
};

export type PlanetItemProps = {
  name: string;
  diameter: string;
  orbital_period: string;
  population: string;
  rotation_period: string;
  surface_water: string;
  gravity: string;
  terrain: string;
  climate: string;
  films: string[];
  residents: string[];
  url: string;
};

export type SpeciesProps = {
  item: SpeciesItemProps;
  isItemFavourite: boolean;
  setIsItemFavourite: (boolean: boolean) => void;
};

export type SpeciesItemProps = {
  name: string;
  average_height: string;
  average_lifespan: string;
  classification: string;
  designation: string;
  eye_colors: string;
  hair_colors: string;
  language: string;
  skin_colors: string;
  films: string[];
  people: string[];
  url: string;
};

export type StarhsipProps = {
  item: StarshipItemProps;
  isItemFavourite: boolean;
  setIsItemFavourite: (boolean: boolean) => void;
};

export type StarshipItemProps = {
  name: string;
  model: string;
  manufacturer: string;
  cargo_capacity: string;
  hyperdrive_rating: string;
  consumables: string;
  cost_in_credits: string;
  crew: string;
  length: string;
  max_atmosphering_speed: string;
  passengers: string;
  starship_class: string;
  films: string[];
  pilots: string[];
  url: string;
};

export type VehicleProps = {
  item: VehicleItemProps;
  isItemFavourite: boolean;
  setIsItemFavourite: (boolean: boolean) => void;
};

export type VehicleItemProps = {
  name: string;
  model: string;
  manufacturer: string;
  cargo_capacity: string;
  consumables: string;
  cost_in_credits: string;
  crew: string;
  length: string;
  max_atmosphering_speed: string;
  passengers: string;
  vehicle_class: string;
  films: string[];
  pilots: string[];
  url: string;
};

export type AccordionProps = {
  item: FilmItemProps &
    PersonItemProps &
    PlanetItemProps &
    SpeciesItemProps &
    StarshipItemProps &
    VehicleItemProps;
  currentRoot: string;
};

export type ModalProps = {
  displayName: String;
  result: () => React.ReactNode;
  item: string[];
};

export type State = {
  roots: {
    isLoading: boolean;
    isBranchedRootLoading: boolean;
    payload: object;
    error: string[];
    currentRoot: StateCurrentRoot;
    branchedRoot: FilmItemProps[] &
      PersonItemProps[] &
      PlanetItemProps[] &
      SpeciesItemProps[] &
      StarshipItemProps[] &
      VehicleItemProps[];
  };
};

type StateCurrentRoot = {
  count: number;
  next: string;
  previous: string | null;
  results: FilmItemProps[] &
    PersonItemProps[] &
    PlanetItemProps[] &
    SpeciesItemProps[] &
    StarshipItemProps[] &
    VehicleItemProps[];
  k: string;
};
