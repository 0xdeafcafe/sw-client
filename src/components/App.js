import React from 'react';
import { Container, Navbar, NavbarBrand } from 'reactstrap';

const App = ({ children }) => (
	<div className={'App'}>
		<Navbar
			color={'light'}
			light
		>
			<NavbarBrand href={'/'}>{'Star Wars Client'}</NavbarBrand>
		</Navbar>

		<Container className={'pt-5'}>
			{children}
		</Container>
	</div>
);

export default App;
