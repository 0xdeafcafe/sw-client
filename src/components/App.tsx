
import React, {ReactNode} from 'react';
import { Container, Navbar, NavbarBrand } from 'reactstrap';

interface AppProps {
	children: ReactNode;
  }

const App: React.FC<AppProps> = ({ children }) => (
	<div className={'App'}>
		<Navbar
			color={'light'}
			light
		>
			<NavbarBrand href={'/'}>{'Star Wars Client'}</NavbarBrand>
		</Navbar>

		<Container className={'pt-5'}>
			{children}
		</Container>
	</div>
);

export default App;
