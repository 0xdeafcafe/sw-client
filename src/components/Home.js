import classnames from 'classnames';
import { fetchRoots } from '../actions';
import sentenceCase from 'sentence-case';
import {
	Nav,
	NavItem,
	NavLink,
	TabContent,
	TabPane,
} from 'reactstrap';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';

const Home = () => {
	const dispatch = useDispatch();
	const [tab, setTab] = useState(void 0);
	const roots = useSelector(state => state.roots);

	useEffect(() => {
		dispatch(fetchRoots());
	}, [dispatch]);

	const keys = Object.keys(roots.payload || {});

	return (
		<div>
			<h1>{'My little Star Wars app ðŸ‘¾'}</h1>

			{roots.payload && (
				<div className={'mt-3'}>
					<Nav tabs>
						{keys.map(k => (
							<NavItem key={k}>
								<NavLink
									className={classnames({ active: tab === k })}
									onClick={() => setTab(k)}
								>
									{sentenceCase(k)}
								</NavLink>
							</NavItem>
						))}
					</Nav>

					<TabContent activeTab={tab}>
						{keys.map(k => (
							<TabPane
								key={k}
								tabId={k}
							>
								<code>{'todo: load resource table'}</code>
							</TabPane>
						))}
					</TabContent>
				</div>
			)}
		</div>
	);
};

export default Home;
