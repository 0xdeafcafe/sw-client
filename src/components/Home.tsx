import classnames from "classnames";
import sentenceCase from "sentence-case";
import styled from "styled-components";
import { fetchIndividualRoot, fetchRoots } from "../actions/actions";
import { Nav, NavItem, NavLink, TabContent, TabPane, Button } from "reactstrap";
import CircularProgress from "@mui/material/CircularProgress";

import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { ItemAccordion } from "./Accordion/Accordion";
import { FilmItemProps, PersonItemProps, PlanetItemProps, SpeciesItemProps, StarshipItemProps, VehicleItemProps } from "./types";

const SpinnerContainer = styled.div`
  display: flex;
  justify-content: center;
  padding: 50px;
`;

const StyledButtonContainer = styled.div`
  display: flex;
  justify-content: flex-end;
`

const Home = () => {
  const dispatch = useDispatch();
  const [tab, setTab] = useState<string | undefined>(void 0);

  const roots = useSelector((state: any) => state.roots);
  const isLoading = useSelector((state: any) => state.roots.isLoading);
  const currentRoot = useSelector(
    (state: any) =>
      (state &&
        state.roots &&
        state.roots.currentRoot &&
        state.roots.currentRoot.k) ||
      []
  );

  const currentRootResults = useSelector(
    (state: any) =>
      (state &&
        state.roots &&
        state.roots.currentRoot &&
        state.roots.currentRoot) ||
      []
  );

  useEffect(() => {
    dispatch(fetchRoots());
  }, [dispatch]);

  const keys = Object.keys(roots.payload || {});

  const handleTabClick = (k: string) => {
    dispatch(fetchIndividualRoot(k));
  };
  return (
    <div>
      <h1>{"My little Star Wars app ðŸ‘¾"}</h1>

      {roots.payload && (
        <div className={"mt-3"}>
          <Nav tabs>
            {keys.map((k) => (
              <NavItem key={k}>
                <NavLink
                  className={classnames({ active: tab === k })}
                  onClick={() => {
                    setTab(k);
                    handleTabClick(k);
                  }}
                >
                  {sentenceCase(k)}
                </NavLink>
              </NavItem>
            ))}
          </Nav>

          {isLoading ? (
            <SpinnerContainer>
              <CircularProgress />
            </SpinnerContainer>
          ) : (
            <>
              <TabContent activeTab={tab}>
                {keys.map((k) => (
                  <TabPane key={k} tabId={k}>
                    {currentRootResults && currentRootResults.results && currentRootResults.results.map((item: FilmItemProps & PersonItemProps & PlanetItemProps & SpeciesItemProps & StarshipItemProps & VehicleItemProps) => (
                      <ItemAccordion currentRoot={currentRoot} item={item} />
                    ))}
                  </TabPane>
                ))}
                <StyledButtonContainer>
                  {currentRootResults.previous &&<Button onClick={() => dispatch(fetchIndividualRoot(currentRootResults.previous))} >Previous</Button>}
                  {currentRootResults.next &&<Button onClick={() => dispatch(fetchIndividualRoot(currentRootResults.next))}>Next</Button>}
                </StyledButtonContainer>
              </TabContent>
            </>
          )}
        </div>
      )}
    </div>
  );
};

export default Home;
