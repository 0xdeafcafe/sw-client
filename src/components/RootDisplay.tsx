import React, {useState, useEffect} from "react";
import { RootPerson } from "./Roots/RootPerson";
import { RootPlanets } from "./Roots/RootPlanets";
import { RootSpecies } from "./Roots/RootSpecies";
import { RootStarships } from "./Roots/RootStarships";
import { RootVehicles } from "./Roots/RootVehicles";
import { RootFilms } from "./Roots/RootFilms";
import { DisplayProps } from "./types";

export const RootDisplay = ({ item, item: {url}, currentRoot }: DisplayProps) => {
  const [isItemFavourite, setIsItemFavourite] = useState(false)

  useEffect(() => {
    const localStorageItems = JSON.parse(localStorage.getItem("session") || '[]')

    if(localStorageItems.includes(url)) {
      setIsItemFavourite(true)
    }
  })

  return(
  <>
    {currentRoot === "people" && <RootPerson setIsItemFavourite={setIsItemFavourite} isItemFavourite={isItemFavourite} item={item} />}
    {currentRoot === "planets" && <RootPlanets setIsItemFavourite={setIsItemFavourite} isItemFavourite={isItemFavourite} item={item} />}
    {currentRoot === "films" && <RootFilms setIsItemFavourite={setIsItemFavourite} isItemFavourite={isItemFavourite} item={item} />}
    {currentRoot === "species" && <RootSpecies setIsItemFavourite={setIsItemFavourite} isItemFavourite={isItemFavourite} item={item} />}
    {currentRoot === "vehicles" && <RootVehicles setIsItemFavourite={setIsItemFavourite} isItemFavourite={isItemFavourite}item={item} />}
    {currentRoot === "starships" && <RootStarships setIsItemFavourite={setIsItemFavourite} isItemFavourite={isItemFavourite} item={item} />}
  </>
);
}