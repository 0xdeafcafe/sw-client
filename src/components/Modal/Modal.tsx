import React, { useState } from "react";
import { Box, styled } from "@mui/system";
import { useDispatch, useSelector } from "react-redux";
import { Button } from "reactstrap";
import { fetchArrayOfItems } from "../../actions/actions";
import ModalUnstyled from "@mui/base/ModalUnstyled";
import CircularProgress from "@mui/material/CircularProgress";
import { ModalProps } from "components/types";


const StyledModal = styled(ModalUnstyled)`
  position: fixed;
  z-index: 1300;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background: white;
`;

const StyledButton = styled(Button)`
  margin-bottom: 5px;
  width: 100%;
`;

const Backdrop = styled("div")`
  z-index: -1;
  position: fixed;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
`;

const SpinnerContainer = styled("div")`
  display: flex;
  justify-content: center;
  padding: 50px;
`;


const style = {
  maxHeight: "600px",
  overflow: "scroll",
  width: 400,
  bgcolor: "background.white",
  border: "2px solid #000",
  p: 2,
  px: 4,
  pb: 3,
};

export default function Modal ({ item, displayName, result }: ModalProps) {
  const dispatch = useDispatch();
  const isBranchedRootLoading = useSelector((state: any) => state.roots.isBranchedRootLoading);
  const [open, setOpen] = useState(false);
  const handleOpen = () => {
    dispatch(fetchArrayOfItems(item));
    setOpen(true);
  };
  const handleClose = () => setOpen(false);
  return (
    <div>
      <StyledButton onClick={handleOpen} type="button">
        {displayName}
      </StyledButton>
      <StyledModal
        aria-describedby="unstyled-modal-description"
        aria-labelledby="unstyled-modal-title"
        BackdropComponent={Backdrop}
        onClose={handleClose}
        open={open}
      >
        <Box sx={style}>
          <h2 id="unstyled-modal-title">{displayName}</h2>
          <p id="unstyled-modal-description">{isBranchedRootLoading ?             
          (<SpinnerContainer>
              <CircularProgress />
            </SpinnerContainer>) : result()}</p>
        </Box>
      </StyledModal>
    </div>
  );
}
