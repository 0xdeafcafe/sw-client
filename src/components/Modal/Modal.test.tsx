import { ModalProps } from "components/types";
import React, { ReactNode } from "react";
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store'
import Modal from "./Modal";
import initialState from "reducers/initial-state";

import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

const mockStore = configureStore();
configure({ adapter: new Adapter() });

const props: ModalProps = {
    displayName: "Species",
    result: function (): ReactNode {
       return (<p>the species name...</p>)
    },
    item: ["https://swapi.dev/api/species/2/"]
}

const store = mockStore(initialState);

it("has the correct text displayed", () => {
    const wrapper = mount(<Provider store = {store}><Modal {...props} /></Provider>)
    expect(wrapper).toMatchSnapshot();
})