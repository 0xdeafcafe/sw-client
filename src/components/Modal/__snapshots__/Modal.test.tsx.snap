// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has the correct text displayed 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Modal
    displayName="Species"
    item={
      Array [
        "https://swapi.dev/api/species/2/",
      ]
    }
    result={[Function]}
  >
    <div>
      <Styled(Button)
        color="secondary"
        onClick={[Function]}
        tag="button"
        type="button"
      >
        <Noop />
        <Button
          className="css-bjdol9"
          color="secondary"
          onClick={[Function]}
          tag="button"
          type="button"
        >
          <button
            aria-label={null}
            className="css-bjdol9 btn btn-secondary"
            onClick={[Function]}
            type="button"
          >
            Species
          </button>
        </Button>
      </Styled(Button)>
      <Styled(ForwardRef(ModalUnstyled))
        BackdropComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "__emotion_base": "div",
            "__emotion_forwardProp": [Function],
            "__emotion_real": [Circular],
            "__emotion_styles": Array [
              "
  z-index: -1;
  position: fixed;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
",
              [Function],
              "",
            ],
            "defaultProps": undefined,
            "displayName": "Styled(div)",
            "render": [Function],
            "withComponent": [Function],
          }
        }
        aria-describedby="unstyled-modal-description"
        aria-labelledby="unstyled-modal-title"
        onClose={[Function]}
        open={false}
      >
        <Noop />
        <ForwardRef(ModalUnstyled)
          BackdropComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "__emotion_base": "div",
              "__emotion_forwardProp": [Function],
              "__emotion_real": [Circular],
              "__emotion_styles": Array [
                "
  z-index: -1;
  position: fixed;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
",
                [Function],
                "",
              ],
              "defaultProps": undefined,
              "displayName": "Styled(div)",
              "render": [Function],
              "withComponent": [Function],
            }
          }
          aria-describedby="unstyled-modal-description"
          aria-labelledby="unstyled-modal-title"
          className="css-497lzk"
          onClose={[Function]}
          open={false}
        />
      </Styled(ForwardRef(ModalUnstyled))>
    </div>
  </Modal>
</Provider>
`;
